File: my_py_pkg/my_py_pkg/number_counter.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.msg import Int64
from example_interfaces.srv import SetBool


class NumberCounterNode(Node):
    def __init__(self):
        super().__init__("number_counter")
        self.counter_ = 0
        self.subscriber_ = self.create_subscription(
            Int64, "number",self.callback_number_counter,10)
        self.publisher_ = self.create_publisher(
            Int64, "number_count", 10)
        self.get_logger().info("Number Counter has been started")
        self.reset_counter_service_ = self.create_service(
            SetBool, "reset_counter", self.callback_reset_counter)
        self.get_logger().info("Reset Counter Service has been started")
        
    def callback_number_counter(self, msg: Int64):
        self.counter_ +=msg.data
        new_msg = Int64()
        new_msg.data = self.counter_
        self.publisher_.publish(new_msg)

    def callback_reset_counter(self, request:SetBool.Request, response:SetBool.Response):
        if request.data:
            self.counter_ = 0
            response.success = True
            response.message = "Counter reset to zero."
        else:
            response.success = False
            response.message = "Counter not reset."
        self.get_logger().info(response.message)
        return response

        
def main(args=None):
     rclpy.init(args=args)
     node = NumberCounterNode()
     rclpy.spin(node)
     rclpy.shutdown()


if __name__ == "__main__":
    main()

File: my_cpp_pkg/src/number_counter.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"
#include "example_interfaces/srv/set_bool.hpp"

using namespace std::placeholders;

class NumberCounterNode : public rclcpp::Node
{
public:

    NumberCounterNode() : rclcpp::Node("number_counter")
    {
        subscriber_ = this-> create_subscription<example_interfaces::msg::Int64>(
            "number",10,
            std::bind(&NumberCounterNode::callbackNumberCounter, this, _1));
        publisher_ = this->create_publisher<example_interfaces::msg::Int64>("number_count", 10);
        service_ = this->create_service<example_interfaces::srv::SetBool>(
            "reset_counter",
            std::bind(&NumberCounterNode::callbackResetCounter, this, _1, _2));
    }

private:
    void callbackNumberCounter(const example_interfaces::msg::Int64::SharedPtr msg)
    {
        number_counter_ += msg->data;
        auto count_msg = example_interfaces::msg::Int64();
        count_msg.data = number_counter_;
        publisher_->publish(count_msg);
    }
    void callbackResetCounter(
        const example_interfaces::srv::SetBool::Request::SharedPtr request,
        example_interfaces::srv::SetBool::Response::SharedPtr response)
    {
        if (request->data) {
            number_counter_ = 0;
            response->success = true;
            response->message = "Counter reset to zero.";
        } else {
            response->success = false;
            response->message = "Counter not reset.";
        }
    }

    rclcpp::Subscription<example_interfaces::msg::Int64>::SharedPtr subscriber_;
    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr publisher_;
    rclcpp::Service<example_interfaces::srv::SetBool>::SharedPtr service_;
    int number_counter_ = 0;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberCounterNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}